### PR TITLE
Support for the Java LDAP client

### PR DESCRIPTION
**Pull request by tobiasstadler**
## What does this PR do?
I added support for instrumenting the Java LDAP client

## Checklist
- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by tobiasstadler** [comment]
@felixbarny Could you please review this PR also?

**Review by felixbarny — APPROVED** [review]

**Comment by felixbarny** [comment]
run elasticsearch-ci/docs

**Comment by tobiasstadler** [comment]
Thank you!

File: apm-agent-plugins/apm-java-ldap-plugin/src/main/java/co/elastic/apm/agent/java_ldap/LdapClientAdvice.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ */
+package co.elastic.apm.agent.java_ldap;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.GlobalTracer;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Outcome;
+import co.elastic.apm.agent.impl.transaction.Span;
+import com.sun.jndi.ldap.Connection;
+import com.sun.jndi.ldap.LdapResult;
```
Comment (felixbarny) [inline apm-agent-plugins/apm-java-ldap-plugin/src/main/java/co/elastic/apm/agent/java_ldap/LdapClientAdvice.java:27]: Seems like IntelliJ has issues with the imports from the `java.naming/com.sun.jndi.ldap` module.


Screen Shot 2022-01-28 at 13 52 13

Applying the suggestion only works until the maven project is reloaded.

I've tried adding
```xml
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <configuration>
                    <compilerArgs>
                        <arg>--add-exports</arg>
                        <arg>java.naming/com.sun.jndi.ldap=ALL-UNNAMED</arg>
                    </compilerArgs>
                </configuration>
            </plugin>
        </plugins>
    </build>
```

But that yields this error: `java: exporting a package from system module java.naming is not allowed with --release`

@tobiasstadler Did you manage to get this work in an IDE?

**Comment by tobiasstadler** [comment]
Can you please try adding `<maven.compiler.source>1.7</maven.compiler.source>` to the properties of the plugin pom?

File: apm-agent-plugins/apm-java-ldap-plugin/src/main/java/co/elastic/apm/agent/java_ldap/LdapClientAdvice.java (hunk 1 above)
Comment (tobiasstadler) [reply apm-agent-plugins/apm-java-ldap-plugin/src/main/java/co/elastic/apm/agent/java_ldap/LdapClientAdvice.java:27]: Can you please try adding <maven.compiler.source>1.7</maven.compiler.source> to the properties of the plugin pom?

Comment (tobiasstadler) [reply apm-agent-plugins/apm-java-ldap-plugin/src/main/java/co/elastic/apm/agent/java_ldap/LdapClientAdvice.java:27]: The problem probably is that the `IntelliJ` profile in the parent pom sets `maven.compiler.target` and `maven.compiler.target` to 11.

Comment (tobiasstadler) [reply apm-agent-plugins/apm-java-ldap-plugin/src/main/java/co/elastic/apm/agent/java_ldap/LdapClientAdvice.java:27]: I just tested it with Java17. It compiles fine but the tests are failing with the error you mentioned.

Comment (tobiasstadler) [reply apm-agent-plugins/apm-java-ldap-plugin/src/main/java/co/elastic/apm/agent/java_ldap/LdapClientAdvice.java:27]: #2430

**Pull request merged**